### PR TITLE
perf(plugin-server): reduce number of `person` lookups in the event pipeline

### PR DESCRIPTION
**Pull request by macobo**
## Problem

~~Depends on https://github.com/PostHog/posthog/pull/10321, do not merge this in!~~

We're doing additional lookups everywhere in the event pipeline for `person` information. We can side-step all that by passing forward information from previous steps inside `PreIngestionEvent`/`IngestionEvent`.

## Changes

- `PreIngestionEvent` now contains an optional `person` field which we return for events with person property updates
- Buffering, event creation logic and action matching now handle `person` field being passed and enhance the payload each step.
- Unify `CachedPersonData` field into a proper duck type of Person.

## How did you test this code?

Unit tests.

### CODE REVIEW

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 1]
```diff
@@ -235,16 +253,14 @@ export class EventsProcessor {
...
-
         const useExternalSchemas = this.clickhouseExternalSchemasEnabled(teamId)
         // proto ingestion is deprecated and we won't support new additions to the schema
         const message = useExternalSchemas
             ? (EventProto.encodeDelimited(EventProto.create(eventPayload)).finish() as Buffer)
             : Buffer.from(
                   JSON.stringify({
                       ...eventPayload,
-                      person_id: personInfo?.uuid,
+                      person_id: eventPersonUuid,
```
Comment (macobo) [inline plugin-server/src/worker/ingestion/process-event.ts:263]: Q to @yakkomajuri - this is changing from undefined -> null if person could not be found, but person properties was previously null. Thoughts on handling bottom types here?

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 2]
```diff
@@ -212,16 +222,24 @@ export class EventsProcessor {
...
+            eventPersonUuid = personInfo.uuid
+            eventPersonProperties = JSON.stringify(personInfo.properties)
+        } else {
+            personInfo = await this.db.getPersonData(teamId, distinctId)
+            if (personInfo) {
+                // For consistency, we'd like events to contain the properties that they set, even if those were changed
+                // before the event is ingested. Thus we fetch the updated properties but override the values with the event's
+                // $set properties if they exist.
+                const latestPersonProperties = personInfo ? personInfo?.properties : {}
+                eventPersonProperties = JSON.stringify({ ...latestPersonProperties, ...(properties.$set || {}) })
```
Comment (macobo) [inline plugin-server/src/worker/ingestion/process-event.ts:241]: cc @yakkomajuri note that this logic here is incorrect and does not handle $set_once properties. Should I file a bug here?

I think I can 'incidentally' solve this once I move processing outside of here.

Also, no unit tests for this function = :(

File: plugin-server/src/worker/ingestion/process-event.ts [hunk 3]
```diff
@@ -212,16 +222,24 @@ export class EventsProcessor {
...
-            // before the event is ingested. Thus we fetch the updated properties but override the values with the event's
-            // $set properties if they exist.
-            const latestPersonProperties = personInfo ? personInfo?.properties : {}
-            eventPersonProperties = JSON.stringify({ ...latestPersonProperties, ...(properties.$set || {}) })
+            eventPersonUuid = personInfo.uuid
+            eventPersonProperties = JSON.stringify(personInfo.properties)
+        } else {
+            personInfo = await this.db.getPersonData(teamId, distinctId)
+            if (personInfo) {
+                // For consistency, we'd like events to contain the properties that they set, even if those were changed
```
Comment (macobo) [inline plugin-server/src/worker/ingestion/process-event.ts:237]: Note: I think we can kill this clause once we move person processing before event processing

**Review by Twixes — APPROVED** [review]
Looks good to me

**Pull request merged**